optimistix/_solver/bisection.py

```diff
@@ -18,6 +20,61 @@ class _BisectionState(eqx.Module, strict=True):
     error: Float[Array, ""]
 
 
+class _ExpansionCarry(eqx.Module, strict=True):
+    lower: Scalar
+    upper: Scalar
+
+
+def _interval_contains_root(carry: _ExpansionCarry,
+                            /,
+                            *,
+                            need_positive: Bool[Array, ""],
+                            expand_upper: Bool[Array, ""],
+                            fn: Fn[Scalar, Scalar, Aux],
+                            args: Any,
+                            ) -> Bool[Array, ""]:
+    new_boundary = jnp.where(expand_upper, carry.upper, carry.lower)
+    carry_val, _ = fn(new_boundary, args)
+    return need_positive ^ (carry_val > 0.0)
+
+
+def _expand_interval(carry: _ExpansionCarry,
+                     /,
+                     *,
+                     expand_upper: Bool[Array, ""],
+                     ) -> _ExpansionCarry:
+    new_domain = 2.0 * (carry.upper - carry.lower)
+    new_lower = jnp.where(expand_upper, carry.upper, carry.lower - new_domain)
+    new_upper = jnp.where(expand_upper, carry.upper + new_domain, carry.lower)
+    return _ExpansionCarry(new_lower, new_upper)
+
+
+def _expand_interval_repeatedly(initial_interval: _ExpansionCarry,
+                                upper_val: Scalar,
+                                lower_val: Scalar,
+                                need_positive: Bool[Array, ""],
+                                *,
+                                fn: Fn[Scalar, Scalar, Any],
+                                args: PyTree,
+                                ) -> _ExpansionCarry:
+    expand_upper = need_positive ^ (upper_val < lower_val)
+    cond_fun = ft.partial(_interval_contains_root,
+                          need_positive=need_positive,
+                          expand_upper=expand_upper,
+                          fn=fn,
+                          args=args)
+    body_fun = ft.partial(_expand_interval, expand_upper=expand_upper)
+    return eqxi.while_loop(cond_fun, body_fun, initial_interval, kind="lax")
```

Comment:
If it's going to be marked with `kind="lax"` then I'd suggest just using `lax.while_loop` directly.

(For context, the main purpose of `eqxi.while_loop` is to be a backpropagatable while loop, which is a thing accomplished using checkpointing. We have two main implementations, `kind="checkpointed"` -- the only to use 99% of the time -- and `kind="bounded"` -- which is an older first attempt. For completeness I also added the ability to fall back to just `kind="lax"`, but that's only really useful if you want to be polymorphic over loop implementations.)

Reply:
Should we switch to checkpointed?  Will this affect our ability to take derivatives through this computation?

Reply:
If we do this right, then I think this shouldn't affect our ability to compute derivatives. The mathematical definition of the gradient shouldn't be affected by our choice of interval.

Reply:
Sounds great.